docs: initial helm charts guide (#209)
open-component-model/ocm-website: content/en/docs/guides/build-deploy-software-via-helm-charts-with-ocm-101.md

1. a public Helm Chart repository
2. local `*.tgz` file
3. local folder which containing Helm Chart files

First we need to download and unpack the Helm Chart for `podinfo`.
This can easily be achieved with the helm CLI:

```shell
helm repo add <repo-name> <helm-chart-repo-url>
helm pull --destination <target-dir> <repo-name/chart-name>
```

`podinfo` example:
```shell
helm repo add podinfo https://stefanprodan.github.io/podinfo
helm pull --destination . podinfo/podinfo
```
The helm chart is then stored in the current working directory as `podinfo-6.6.0.tgz` and can be referenced as path from there in the `component-constructor.yaml` file (see below).

Unpack `podinfo-6.6.0.tgz` to simulate the process as if this helm chart is our own and is not downloaded from a public repository.
```shell
tar -czf podinfo-6.6.0.tgz podinfo
```

### Input Specification

The corresponding input file for building the component version (`component-constructor.yaml`) will then look like this:

```yaml
components:
- name: ${COMPONENT_NAME_PREFIX}/podinfo
  version: ${PODINFO_VERSION}
  provider:
    name: ${PROVIDER}
  resources:
  - name: helm-chart-external
    type: helmChart    

>     relation: external

Reviewer: Is it considered a best practive to specify the `relation` field explicitly?
It is set automatically by the ocm cli.

Author: @morri-son thx, I will remove this form the guide.

    access:
      type: helm
      helmChart: podinfo:${PODINFO_CHART_VERSION}
      helmRepository: https://stefanprodan.github.io/podinfo
  - name: helm-chart-local-tgz
    type: helmChart
    relation: local    
    input:      
      type: helm

>       path: podinfo-${PODINFO_CHART_VERSION}.tgz

Reviewer: It seems that you tried to show all possibilities to define a helmChart resource.
Why not adding the external helm chart as input by its helm repo location. By that one can avoid to download the chart first but let ocm handle the download.

```yaml
- name: helm-chart-local-from-repo
  type: helmChart
  input:
    type: helm
    helmRepository: https://stefanprodan.github.io/podinfo
    path: podinfo
    version: ${PODINFO_CHART_VERSION}
```

  - name: helm-chart-local-folder
    type: helmChart
    relation: local    
    input:   
      type: dir

>       path: ./podinfo/ 

Reviewer: It seems that you tried to show all possibilities to define a `helmChart` resource.
Why not adding the external helm chart as `input` by its helm repo location. By that one can avoid to download the chart first but let ocm handle the download.

```yaml
  - name: helm-chart-local-from-repo
    type: helmChart
    input:
      type: helm
      helmRepository: https://stefanprodan.github.io/podinfo
      path: podinfo
      version: ${PODINFO_CHART_VERSION}
```

Reviewer: But you described downloading the helm chart first. From your description it looks like downloading the helm chart first with helm is a requirement when it isn't.
When using an external helm chart I would prefer to reference the external helm repository directly in the ocm input file.
Besides not needing the helm binary locally is, that one must only specify credentials for the ocm tool and not for several other tools (docker, helm etc.).

I see the input type `dir` as more useful when a local helm chart (stored in the same Git repository) shall be added to the component.

Author: @reshnm correct. It's downloaded to demonstrate the different possibilities of packing a Helm Chart as OCM resource.  I avoided the extra step for the guide reader of creating a new helm chart.

As written in the guide:

`To demonstrate No. 2+3 we first we need to download and unpack the Helm Chart for `podinfo`. In a real application, this would be your own Helm Charts.`

Reviewer 2: this is indeed a bit misleading. I get the point of showing all possible ways to define a Helm chart resource, but as Rene mentioned, it looks a bit like a prerequisite even when mentioning the download. And for local helm charts dir makes more sense.

Author: @morri-son / @reshnm  thx for pointing that out in detail. Now I understand it much better. I will add an explanation later on.

Author: Updated to:

Reference a Helm Chart from:

1. a public Helm Chart repository
2. local `*.tgz` file
3. local folder containing a Helm Chart file

To demonstrate No. 2. and 3. we need a Helm Chart on our local machine. For the sake of the this simplified guide, we download and unpack a already existing open source Helm Chart for `podinfo`. In a real world application, this would be your own Helm Chart. You will most likely already store your own Helm Charts within a `git` repository and leverage CI/CD Pipeline to build `*.tgz` Helm Chart files in order to push to your Helm Chart Repository.

>   - name: image

Reviewer: Here there is no distinction between `external` and `local`.
But you could also create it as a local resource:

```yaml
- name: local-image
    type : ociImage
    input:
       type: ociArtifact
       path: ghcr.io/stefanprodan/podinfo:${PODINFO_VERSION} 
       repository: ocm/ocm.software/podinfo/image
```

```

```shell
ocm transfer commontransportarchive --copy-resources --enforce --overwrite ./ocm-hello-world OCIRegistry::ghcr.io/stb1337/ocm-hello-world-2
transferring component "ocm.software/podinfo"...
  transferring version "ocm.software/podinfo:6.6.0"...
    version "ocm.software/podinfo:6.6.0" already present -> but requires resource transport
  ...resource 0 helm-chart-external[helmChart] (copy)
  ...resource 1 helm-chart-local-tgz[helmChart](ocm.software/podinfo/podinfo:6.6.0) (already present)
  ...resource 2 helm-chart-local-folder[helmChart] (already present)  
  ...resource 3 image[ociImage](stefanprodan/podinfo:6.6.0) (copy)
  ...adding component version...
```

Note: Be careful with the `-f` or `--overwrite` flag. This will replace existing component
versions in the OCI registry. During development it is useful being able to overwrite
existing component versions until something is ready for release. **For released versions
you should never use this flag**! Released component versions should be immutable and
should never be overwritten. They serve as source of truth for what the release is made of
und should never be changed.

### Package Visibility 

> Change Github Package Visibility to 'plublic':

Reviewer: ```suggestion
Change Github Package Visibility to 'public':
```

---
title: "Build & Deploy Infrastructure via Helm Charts with OCM" 
description: ""
lead: ""
date: 2024-03-19T10:36:48+01:00
lastmod: 2024-03-19T10:36:48+01:00
draft: false
images: []
menu:
  docs:
    parent: ""
    identifier: "build-deploy-software-via-helm-charts-with-ocm"
weight: 101
toc: true
---

## Introduction

> Let's illustrate a very simple "Hello World" example application and show how to leverage OCM to build and deploy a Helm Chart and OCI Image on a local `kind` k8s cluster.

Reviewer 2: ```suggestion
Let's illustrate a very simple "Hello World" example application and show how to leverage OCM to build an application component containing a Helm Chart and an OCI Image and deploy it to a local `kind` k8s cluster.
```

> The topic of `ocm` [`localization`](https://ocm.software/docs/guides/complex-component-structure-deployment/#localization) and [`configuration`](https://ocm.software/docs/guides/complex-component-structure-deployment/#configuration) is NOT part of this very simple example and is part of other in deepth guides.

Reviewer 2: ```suggestion
The topics `ocm` [`localization`](https://ocm.software/docs/guides/complex-component-structure-deployment/#localization) and [`configuration`](https://ocm.software/docs/guides/complex-component-structure-deployment/#configuration) are NOT part of this very simple example and is part of other deep-dive guides.
```

> As base we use the `podinfo` application.

Reviewer 2: ```suggestion
As base we use the `podinfo` application from Stefan Prodan's [Github repo](https://github.com/stefanprodan/podinfo).
```

> All files can be found [here](https://github.com/stb1337/ocm-hello-world-v1).

Reviewer 2: I propose to re-activate the https://github.com/open-component-model/ocm-examples repository and move all files required for examples linked on the website to it. The podinfo component file(s) may end up [here](https://github.com/open-component-model/ocm-examples/tree/main/components) and all other required files under the kubernetes folder.

Author: I can move everything there, but therefore this repository needs to be public again. 
Additionally, I would add everything into `[ocm-examples](https://github.com/open-component-model/ocm-examples/tree/main)/[components](https://github.com/open-component-model/ocm-examples/tree/main/components)/guide-walkthrough-helm-chart` and `[ocm-examples](https://github.com/open-component-model/ocm-examples/tree/main)/[kubernetes](https://github.com/open-component-model/ocm-examples/tree/main/kubernetes)/guide-walkthrough-helm-chart` ?

Author: done, I have moved all files to ocm-example -> https://github.com/open-component-model/ocm-examples/pull/25

Author: @morri-son updated references to ocm-example repository. Please resolve conversation and approve PR.

At the end of the tutorial the application will consist of the following components:

> * `podinfo` as our business application

Reviewer 2: since it is just one component I propose to skip the list and the following sentence in line 29 in favour of a single, combined sentence describing the result, an OCM component and its two versions, containing the business application podinfo. I'm not sure if mentioning the six OCI images makes sense here.

Reviewer 2: I did a change in your main branch to this section. Please check if this make sense.

Following the OCM guidelines we will end-up with two component versions and six OCM OCI images.

> For building composed components the ["All-in-one"](../getting-started-with-ocm#all-in-one)

Reviewer 2: ```suggestion
For building composed components the ["all-in-one"](../getting-started-with-ocm#all-in-one)
```

mechanism becomes handy.

### Requirements

* [OCM command line tool](https://github.com/open-component-model/ocm)
* [kubectl](https://kubernetes.io/docs/reference/kubectl/)
* [git](https://git-scm.com/downloads)
* [kind](https://kind.sigs.k8s.io/docs/user/quick-start/#installation)
* [flux](https://fluxcd.io/flux/installation/#install-the-flux-cli)

> ## Building OCM Software Artifact

Reviewer 2: ```suggestion
## Building the Application Component using OCM
```

First we build an OCM component which contains Helm Charts in different kind of formats. This 101 guide explains all possible formats,
but in reality you'll just pick the one most appropiate to you.

### Prepare Helm Charts

> We are leveraging Kubernetes deployments which often use helm charts. The OCM specification supports helm charts as

Reviewer 2: ```suggestion
We are leveraging Kubernetes deployments which often use Helm charts. The OCM specification supports Helm charts as
```

> an artifact type. For this simple example, we will re-use already existing open source community helm charts.

Reviewer 2: ```suggestion
an artifact type. For this simple example, we will re-use already existing open source community Helm charts.
```

> The OCM CLI supports referencing helm charts being stored in an OCI registry. However most

Reviewer 2: ```suggestion
The OCM CLI supports referencing Helm charts being stored in an OCI registry. However most
```

publicly available helm charts currently are available from helm chart repositories and

> not from OCI registries. Therefore helm charts can be embedded in the `component archive` in three different ways.

Reviewer 2: ```suggestion
not from OCI registries. Therefore Helm charts can be embedded in the `component archive` in three different ways.
```

Reference a Helm Chart from:

1. a public Helm Chart repository
2. local `*.tgz` file
3. local folder containing a Helm Chart file

> To demonstrate No. 2. and 3. we need a Helm Chart on our local machine. For the sake of the this simplified guide, we download and unpack a already existing open source Helm Chart for `podinfo`. In a real world application, this would be your own Helm Chart. You will most likely already store your own Helm Charts within a `git` repository and leverage CI/CD Pipeline to build `*.tgz` Helm Chart files in order to push to your Helm Chart Repository.

Reviewer 2: ```suggestion
To demonstrate No. 2. and 3. we need a Helm chart on our local machine. For the sake of the this simplified guide, we download and unpack an already existing open source Helm chart for `podinfo`. In a real world application, this would be your own Helm chart. You will most likely already store your own Helm charts within a `git` repository and leverage a CI/CD pipeline to build `*.tgz` Helm chart files in order to push to your Helm chart repository.
```
